node/src/braid/mod.rs

```diff
@@ -17,99 +19,144 @@ pub enum AddBeadStatus {
+    pub(crate) tips: HashSet<usize>,
     pub(crate) cohorts: Vec<Cohort>,
-    pub(crate) orphan_beads: Vec<Bead>,
-    pub(crate) genesis_beads: HashSet<BeadHash>,
+    pub(crate) orphan_beads: Vec<usize>,
+    pub(crate) genesis_beads: HashSet<usize>,
+    pub(crate) bead_index_mapping: HashMap<BeadHash, usize>,
 }
 
 impl Braid {
-    // All public funtions go here!
-    pub fn new(genesis_beads: HashSet<BeadHash>) -> Self {
-        Braid {
-            beads: genesis_beads.clone(),
-            tips: genesis_beads.clone(),
-            cohorts: vec![Cohort(genesis_beads.clone())],
-            orphan_beads: Vec::new(),
-            genesis_beads: genesis_beads,
-        }
-    }
+    ///Initializing the Braid object for keeping track of current state of Braid
+    pub fn new(genesis_beads: HashSet<Bead>) -> Self {
+        let mut beads = Vec::new();
+        let mut bead_indices = HashSet::new();
+        let mut bead_index_mapping = HashMap::new();
 
-    pub fn add_bead(&mut self, bead: Bead) -> AddBeadStatus {
-        if self.beads.contains(&bead.block_header.block_hash()) {
-            return AddBeadStatus::DagAlreadyContainsBead;
+        for (index, bead) in genesis_beads.into_iter().enumerate() {
+            beads.push(bead.clone());
+            bead_indices.insert(index);
+            bead_index_mapping.insert(bead.block_header.block_hash(), index);
         }
 
-        self.cohorts = self.calculate_cohorts();
-
-        AddBeadStatus::BeadAdded
+        Braid {
+            beads,
+            tips: bead_indices.clone(),
+            cohorts: vec![Cohort(bead_indices.clone())],
+            orphan_beads: Vec::new(),
+            genesis_beads: bead_indices,
+            bead_index_mapping,
+        }
     }
 }
-
+#[allow(unused)]
 impl Braid {
     /// Attempts to extend the braid with the given bead.
     /// Returns true if the bead successfully extended the braid, false otherwise.
-    fn extend(&mut self, bead: &Bead) -> bool {
+    pub fn extend(&mut self, bead: &Bead) -> bool {
         // No parents: bad block
         if bead.committed_metadata.parents.is_empty() {
             return false;
         }
         // Don't have all parents
-        if !bead
-            .committed_metadata
-            .parents
-            .iter()
-            .all(|p| self.beads.contains(p))
-        {
-            return false;
+        for parent_hash in &bead.committed_metadata.parents {
+            // Check if we already have this parent in our beads
+            let parent_exists = self
+                .beads
+                .iter()
+                .any(|b| b.block_header.block_hash() == *parent_hash);
+
+            if !parent_exists {
+                // Try to retrieve the parent
+                if let Some(retrieved_bead) = retrieve_bead(*parent_hash) {
+                    self.extend(&retrieved_bead);
+                } else {
+                    // Parent not found and can't be retrieved
```

Comment:
This is a silent failure. Here we need to be adding this bead to `orphan_beads`.

At the end of `extend()` we need to be processing anything in `orphan_beads` in case a missing parent was just added.